.semgrep.yml

```diff
@@ -0,0 +1,32 @@
+rules:
+  - id: use-pure-path
+    patterns:
+      - pattern: pathlib.Path
+    message: >-
+      Don't use pathlib.Path, use pathlib.PurePath instead: file IO should happen via the engine, not Python
+    languages:
+      - python
+    severity: ERROR
+    paths:
+      # this only matters in code that runs as part of `pants ...`
+      exclude:
+        - "*_test.py"
+      include:
+        - src/python/pants/
+        - pants-plugins/
```

Comment:
Truly it's just rule code. Could our pattern only match inside a function decorated with `@rule`? Is it that configurable?

Reply:
Yeah, one can constrain it to only apply to `@rule`-annotated functions:  using `pattern-inside`: https://semgrep.dev/playground/s/ejQd

```yaml
rules:
  - id: use-pure-path
    patterns:
      - pattern-inside: |
          @rule
          def $FUNC(...):
            ...
      - pattern: pathlib.Path
    message: >-
      Don't use pathlib.Path, use pathlib.PurePath instead: file IO should happen via the engine, not Python
    languages:
      - python
    severity: ERROR
```

```python
from pathlib import Path

@rule
def foo():
    Path("inside")

def bar():
    Path("outside")

Path("also outside")
```

That'll only flag `Path("inside")`. That said, for Pants' actual use case, I think we can have helper functions/types that aren't `@rule`s that still should use `PurePath`, not `Path`.

(If you're curious to experiment, their docs + the playground are pretty useful https://semgrep.dev/docs/writing-rules/overview/ )

Reply:
Fair point! What I'm thinking is we can use it for automation, and maybe for the synchronous parts of Pants.

But if the sync parts don't use it today, then they probably won't be affected by this 😅